inc/Classes/PasswordHash.php

```diff
@@ -0,0 +1,204 @@
+            } else {
+                throw new \Exception('Unsupported hash algorithm configured: ' . $algo_cfg['algorithm']);
+            }
+        }
+        return hash_equals($info['hash'], $newhash);
+    }
+
+    public static function needsRehash($hash)
+    {
+        global $cfg;
+
+        $systemConfig = self::getAlgoCfg();
+
+        $hashProperties = self::getInfo($hash);
+        // rehashing needed if algorithm does not match or amount of iterations is different than system setting
+        // I was thinking about keeping higher iteration counts, but this would prevent any recovery from setting value too high by mistake
+        return ($hashProperties['hash'] != $systemConfig['hash']) || ($hashProperties['iterations'] != $systemConfig['iterations']);
+    }
+
+    private static function getInfo($hash)
+    {
+        if ($hash[0] === '$') {
+            $parts = explode('$', $hash);
+            if (count($parts) !== 5) {
+                throw new Exception('Unexpected hash format');
+            }
+            list($dummy, $algo, $iterations, $b64salt, $b64hash) = $parts;
+
+            if (!is_numeric($iterations) || $iterations < 1) {
+                throw new Exception('Unexpected iterations value');
+            }
+            $rawsalt = base64_decode($b64salt, true);
+            $rawhash = base64_decode($b64hash, true);
+            if ($rawsalt === false || $rawhash === false) {
+                throw new Exception('Unexpected base64_decode error');
+            }
+
+            return array('algorithm' => $algo, 'iterations' => $iterations, 'hash' => $rawhash, 'salt' => $rawsalt);
+        } else {
+            return array('algorithm' => 'md5', 'iterations' => 0, 'hash' => $hash, 'salt' => '');
+        }
+    }
+
+    private static function getAlgo()
+    {
+        global $cfg;
+
+        if (array_key_exists('password_hash_algorithm', $cfg)) {
+            $pwhash_algo = $cfg['password_hash_algorithm'];
+        } else {
+            $pwhash_algo = 'md5';
+        }
+
+        if ($pwhash_algo === 'default') {
+            $pwhash_algo = 'pbkdf2-sha512';
+        }
+
+        return $pwhash_algo;
+    }
+
+    /**
+     * Returns the default config for the algorithm provided
+     * Also returns a default config if selected algorithm has no config
+     * If selected algorithm is not supported on the system it is being overwritten with standard config
+     * @param string $algo The algorithm to receive the config for
+     * @return mixed[] Array of configuration options, currently only algorithm and iterations
+     */
+    public static function getDefaultAlgorithmConfig(string $algo = 'default') : array
+    {
+        if (PasswordHash::isAlgorithmSupported($algo)) {
+            // only consider Algorithm further if supported on the system
+            if (array_key_exists($algo, PasswordHash::PBKDF2_DEFAULT_CONFIG)){
```

Comment:
`$algo` is type string and we try to check if an array key exists. This can't work or am I missing something?